### PR TITLE
Update cobra to 1.2.1

### LINKED ISSUE #1096: autocomplete for `kubectl` version `1.22.0` does not work with `exec -it` option
**What happened**:
autocomplete for `kubectl` version `1.22.0` does not work with `exec -it` option:
- `kubectl exec -it <tab_to_see_pods>` - does not show `pods`
- `kubectl exec <tab_to_see_pods> -it -- bash` - shows `pods`

**What you expected to happen**:
autocomplete for `kubectl` version `1.21.3` works fine with `exec -it` option:
- `k exec -it <tab_to_see_pods>` - shows pods list

**How to reproduce it (as minimally and precisely as possible)**:

1. Install `kubectl` version `1.22.0` and try:
```
kubectl exec -it <tab_to_see_pods>
```
2. Install `kubectl` version `1.21.3` and try:
```
kubectl exec -it <tab_to_see_pods>
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes client and server versions (use `kubectl version`):

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.0", GitCommit:"c2b5237ccd9c0f1d600d3072634ca66cefdf272f", GitTreeState:"clean", BuildDate:"2021-08-04T18:03:20Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.7-eks-d88609", GitCommit:"d886092805d5cc3a47ed5cf0c43de38ce442dfcb", GitTreeState:"clean", BuildDate:"2021-07-31T00:29:12Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.22) and server (1.20) exceeds the supported minor version skew of +/-1
```
- Cloud provider or hardware configuration:
`AWS/GCP/Baremetal`
- OS (e.g: `cat /etc/os-release`):
```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

### PR DESCRIPTION
**Pull request by eddiezane**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR upgrades our version of [cobra](https://github.com/spf13/cobra) to [1.2.1](https://github.com/spf13/cobra/releases/tag/v1.2.1). [1.2.0](https://github.com/spf13/cobra/releases/tag/v1.2.0) overhauled completions.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1096

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by marckhouzam** [comment]
👍
Updating to the latest Cobra release will avoid a small regression introduced when kubectl moved to Go completions: https://github.com/kubernetes/kubernetes/pull/96087#issuecomment-868806797

File: vendor/github.com/spf13/cobra/bash_completions.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ Please refer to [Shell Completions](shell_completions.md) for details.
 
 For backward compatibility, Cobra still supports its legacy dynamic completion solution (described below).  Unlike the `ValidArgsFunction` solution, the legacy solution will only work for Bash shell-completion and not for other shells. This legacy solution can be used along-side `ValidArgsFunction` and `RegisterFlagCompletionFunc()`, as long as both solutions are not used for the same command.  This provides a path to gradually migrate from the legacy solution to the new solution.
 
+**Note**: Cobra's default `completion` command uses bash completion V2.  If you are currently using Cobra's legacy dynamic completion solution, you should not use the default `completion` command but continue using your own.
```
Comment (rothgar) [inline vendor/github.com/spf13/cobra/bash_completions.md:9]: Can you explain what this means? Maybe not in the doc but I don't know what it means and trying to understand.

Comment (marckhouzam) [reply vendor/github.com/spf13/cobra/bash_completions.md:9]: Cobra 1.2 automatically provides a `completion` command for programs that don't provide one themselves.  This greatly reduces the barrier-to-entry for providing completion support.  This "default" completion command uses the new bash completion v2 logic, also introduced in Cobra 1.2.  The v2 bash completion logic only supports custom completions written in Go (which kubectl started using through #96087 ).

But in any case, kubectl is not affected as it already defines its own `completion` command which still points to the bash completion v1 logic.

Please refer to https://github.com/spf13/cobra/blob/master/shell_completions.md#bash-completion-v2

File: staging/src/k8s.io/controller-manager/go.mod [hunk 2]
```diff
@@ -18,6 +18,7 @@ require (
 )
 
 replace (
+	github.com/spf13/viper => github.com/spf13/viper v1.7.0
```
Comment (BenTheElder) [inline staging/src/k8s.io/controller-manager/go.mod:21]: I don't think this is expected. You probably need to run the dependency lint script after fixing the other issues

**Review by BenTheElder — COMMENTED** [review]
the failures in the dependencies presubmit need resolving, looks like other stray changes generated due to that.

File: staging/src/k8s.io/controller-manager/go.mod (hunk 2 above)
Comment (eddiezane) [reply staging/src/k8s.io/controller-manager/go.mod:21]: > I don't think this is expected. You probably need to run the dependency lint script after fixing the other issues

I've been running the scripts for the past few hours 😅. This is going to be a bit messy due to the viper upgrade.

**Comment by eddiezane** [comment]
There was a breaking change in the newer version of mock https://github.com/golang/mock/issues/571.

That should be fixed when https://github.com/kubernetes/kubernetes/pull/102913 is merged.

cc @pacoxu

**Comment by pacoxu** [comment]
#102913 is merged.
Currently, `github.com/spf13/cobra` is v1.1.3.

Updating cobra to 1.2.1 will prevent the 1.2.0 problem. I think we can update it if there is no side effect.

**Comment by eddiezane** [comment]
This is ready for review.

The added transitive deps are:

| dep | license |
| --- | --- |
| github.com/kr/fs | BSD-3-Clause License |
| github.com/pkg/sftp | BSD-2-Clause License |

/assign @liggitt 

/remove-area apiserver
/remove-area cloudprovider
/remove-area code-generation
/remove-sig api-machinery
/remove-sig auth
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/triage accepted
/priority backlog

**Comment by eddiezane** [comment]
Can confirm this fixes https://github.com/kubernetes/kubectl/issues/1096.

**Comment by pacoxu** [comment]
```
@@ -1,6 +1,6 @@
 {
 	"directDependencies": 161,
-	"transitiveDependencies": 303,
-	"totalDependencies": 360,
-	"maxDepthOfDependencies": 27
+	"transitiveDependencies": 300,
+	"totalDependencies": 357,
+	"maxDepthOfDependencies": 29
 }

```
Fewer dependencies, but more max depth.
Overall, lgtm.

**Comment by liggitt** [comment]
/approve
/lgtm

I agree this is a net gain, though the cobra → viper → afero subgraph is tending to pull in additional transitive dependencies with each bump I'd like to avoid adding to the graph ... could be worth looking into isolating / making those optional in the future

**Pull request merged**